app/main.py

```diff
@@ -1,6 +1,23 @@
-from typing import Callable
+from typing import Callable, Any
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cached_data = {}
+    immutable_type = (int, float, str, tuple, bool, type(None))
+
+    def argument_based_cacher(*args) -> Any:
+        nonlocal cached_data
```

Comment:
you're modifying the dictionary that cached_data points to (by adding keys and values). Since you're not re-assigning the variable, you don't need nonlocal.